lib/pbench/agent/results.py

```diff
@@ -122,45 +176,137 @@ def make_result_tb(self) -> str:
         with mdlog_name.open("w") as fp:
             mdlog.write(fp)
 
+        # FIXME: /bin/cp pbench.log ${pbench_run_name}/
+
         tarball = self.target_dir / f"{pbench_run_name}.tar.xz"
+        e_file = self.target_dir / f"{pbench_run_name}.tar.err"
+        args = [self.tar_path, "--create", "--force-local"]
+        if single_threaded:
+            args.append("--xz")
+        args.append(pbench_run_name)
         try:
-            with tarfile.open(tarball, mode="x:xz") as tar:
-                for f in self.result_dir.rglob("*"):
-                    tar.add(os.path.realpath(f))
-        except tarfile.TarError:
-            self.logger.error(
-                "Tar ball creation failed for {}, skipping", self.result_dir
-            )
+            # Invoke tar directly for efficiency.
+            if single_threaded:
+                with tarball.open("w") as ofp, e_file.open("w") as efp:
+                    tar_proc = subprocess.Popen(
+                        args,
+                        cwd=str(self.result_dir.parent),
+                        stdin=None,
+                        stdout=ofp,
+                        stderr=efp,
+                    )
+                    tar_proc.wait()
+                xz_proc = None
+            else:
+                with tarball.open("w") as ofp, e_file.open("w") as efp:
+                    xz_proc = subprocess.Popen(
+                        [self.xz_path, "-T0"],
+                        cwd=str(self.target_dir),
+                        stdin=subprocess.PIPE,
+                        stdout=ofp,
+                        stderr=efp,
+                    )
+                    tar_proc = subprocess.Popen(
+                        args,
+                        cwd=str(self.result_dir.parent),
+                        stdin=None,
+                        stdout=xz_proc.stdin,
+                        stderr=efp,
+                    )
+                    tar_proc.wait()
+                    # Now that the `tar` command has exited, we close the
+                    # `stdin` of the `xz` command to shut it down.
+                    xz_proc.stdin.close()
+                    xz_proc.wait()
```

Comment:
We are using `xz -T0` to parallelize the compression.  This has made a significant performance difference for folks.  This is what `pbench-move-results` does today.

Reply:
OK, so the reason for invoking `xz` separately is to allow us to specify the `-T0` switch, which enables `xz` _itself_ to go multithreaded (which, I'm guessing from your comment, doesn't happen when it is invoked implicitly by `tar`), and it has nothing otherwise to do with the fact that our code runs `xz` separately from `tar`.  Fair enough.

It might be handy to mention that stuff in a code comment.